chore: adds queue to the worker service pattern

xar-tol opened the pull request:
Adds the option for queues to the base worker structure.

Addresses #2550

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

xar-tol committed:
chore: adds worker to template

Adds subscribe data structures to template workload.
Adds template function for generating subscription JSON that lists out subscribed topics, their services, and any related queue they may have.

xar-tol committed:
adds validation/converters for template

adds validations, excluding init/deploy checks for another pr

xar-tol committed:
adds queue to worker

adds queue ds
adds queue validators/converters

xar-tol committed:
sns->sqs pattern

xar-tol committed:
adjusts formatting, bool

xar-tol committed:
adjusts arn to uri

xar-tol committed:
adjusts URI retrieval method

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -474,16 +477,31 @@ func validateTopicSubscription(ts manifest.TopicSubscription, validTopicARNs []s
+		topicName := arn.Resource
 
-		if topicName[2] == ts.Service && topicName[3] == ts.Name {
+		if strings.Contains(topicName, ts.Service) && strings.Contains(topicName, ts.Name) {
```
Are there any edge cases where a malicious user could do something like:
```yaml
subscribe:
  topics:
    - service: eviltopic
      name: worker-events
```
and pass this validation? For example, if `worker` exposes a topic called `events` we'll check the topic subscription against `app-env-worker-events-eviltopic` and this validation will pass. But this would let someone subscribe to the wrong topic. 

Can we pass the app and env into this validator and construct the entire topic name, then check that the constructed `app-env-svc-name` set matches arn.Resource exactly?

bvtujo commented on internal/pkg/manifest/worker_svc.go:
```diff
@@ -38,18 +45,76 @@ type WorkerServiceConfig struct {
+}
+
+// SQSQueue represents the configurable options for setting up a SQS Queue.
+type SQSQueue struct {
+	Name       *string          `yaml:"name"`
+	Retention  *string          `yaml:"retention"`
+	Delay      *string          `yaml:"delay"`
+	Timeout    *string          `yaml:"timeout"`
+	KMS        *bool            `yaml:"kms"`
+	DeadLetter *DeadLetterQueue `yaml:"dead_letter"`
+	FIFO       *FIFOOrBool      `yaml:"fifo"`
+}
```
let's use `*time.Duration` to represent these. This way customers can unmarshal anything from `3600s` to `4d6h` and the duration package will parse it correctly. 

Check out `convertHealthCheck` for examples of unmarshaling and dealing with `duration` objects correctly.
```suggestion
type SQSQueue struct {
	Name       *string          `yaml:"name"`
	Retention  *time.Duration          `yaml:"retention"`
	Delay      *time.Duration          `yaml:"delay"`
	Timeout    *time.Duration          `yaml:"timeout"`
	KMS        *bool            `yaml:"kms"`
	DeadLetter *DeadLetterQueue `yaml:"dead_letter"`
	FIFO       *FIFOOrBool      `yaml:"fifo"`
}
```

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -630,32 +630,124 @@ func convertTopic(t manifest.Topic, accountID, partition, region, app, env, svc
+	if q == nil {
+		return nil, nil
+	}
+	retention, err := convertTime(q.Retention, 0, 1209600)
```
instead of passing in the floor and ceiling for all of these, can we write helper functions like so and use a `time.Duration` for the input?  

```golang
const (
    retentionMinValueSeconds = 0
    retentionMaxValueSeconds = 1209600
)
...
	retention, err := convertRetention(q.Retention)
...
func convertRetention(d *time.Duration) (*int, error) {
    return convertTime(d, retentionMinValueSeconds, retentionMaxValueSeconds)
}
```

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -630,32 +630,124 @@ func convertTopic(t manifest.Topic, accountID, partition, region, app, env, svc
+		return nil, nil
+	}
+
+	tm, err := time.ParseDuration(aws.StringValue(t))
```
If we use the suggestion above, unmarshaling directly from the manifest as a `time.Duration`, we should be able to get rid of this call.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers_test.go:
```diff
@@ -1626,7 +1665,7 @@ func Test_convertSubscribe(t *testing.T) {
 		"topic not allowed": {
 			inSubscribe: &manifest.SubscribeConfig{
-				Topics: &[]manifest.TopicSubscription{
+				Topics: []manifest.TopicSubscription{
```
Can we add a separate unit test validating that a maliciously constructed input fails?

Maliciously constructed meaning it contains the correct elements (service and topic name) with extra info which causes it to point to a different topic.

bvtujo commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -28,6 +30,7 @@ var (
 	errNoSourceVolume           = errors.New("`source_volume` cannot be empty")
 	errEmptyEFSConfig           = errors.New("bad EFS configuration: `efs` cannot be empty")
 	errMissingPublishTopicField = errors.New("field `publish.topics[].name` cannot be empty")
+	errDeadLetterQueueTries     = errors.New("DeadLetter `tries` field cannot exceed 1000")
```
Can we constantize this max value, along with the other bounds, and create error messages that include those constants using `fmt.Errorf()`?

bvtujo reviewed (commented):
This is looking really good. Such complex logic and you've handled it quite well. 

I have some comments about how we validate the sns topic subscriptions and unmarshal the durations (retention, delay, timeout), but otherwise this looks good and I can't wait to see the template changes.

xar-tol committed:
adjusts time to duration, adds constants

xar-tol committed:
adjusts error message for delay, timeout, and retention

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -50,21 +52,27 @@ var (
 	trailingPunctRegExp = regexp.MustCompile(`[\-\.]$`)            // Check for trailing dash or dot.
 )
 
+// Options for SQS Queues
```
heeheehee
```suggestion
// Options for SQS Queues.
```

Lou1415926 commented on internal/pkg/manifest/worker_svc.go:
```diff
@@ -38,18 +46,76 @@ type WorkerServiceConfig struct {
 // SubscribeConfig represents the configurable options for setting up subscriptions.
 type SubscribeConfig struct {
-	Topics *[]TopicSubscription `yaml:"topics"`
+	Topics []TopicSubscription `yaml:"topics"`
```
Can we add some test to see if `Topics` can be properly overridden? Ideally we should include test cases where the original `Topics` is `nil` and the environment `Topics` is non-nil, and vice versa; as well as test cases where the non-nil `Topics` gets overridden by empty environment `Topics`, and vice versa.

Lou1415926 commented on internal/pkg/manifest/worker_svc.go:
```diff
@@ -38,18 +46,76 @@ type WorkerServiceConfig struct {
+// DeadLetterQueue represents the configurable options for setting up a Dead-Letter Queue.
+type DeadLetterQueue struct {
+	ID    *string `yaml:"queue_id"`
+	Tries uint16  `yaml:"tries"`
```
Should this be `*uint16`? I think the override probably doesn't function correctly if it's `uint16`. For example, it might get overriden as `0` if the field is not provided in environment.

Lou1415926 commented on internal/pkg/manifest/worker_svc_test.go:
```diff
@@ -746,3 +746,56 @@ func TestWorkerSvc_ApplyEnv_CountOverrides(t *testing.T) {
+		want     testFIFO
+		wantErr  string
+	}{
+		"fifo specified": {
```
We should also add a test case that shows how we error out if `fifo` gets an invalid input (neither `fifo` nor `enabled`). This will help us maintain and understand the behaviors & catch any future errors if the codebase is changed.

Lou1415926 commented on internal/pkg/template/workload.go:
```diff
@@ -43,9 +43,8 @@ const (
-	snsArnPattern     = "arn:%s:sns:%s:%s:%s-%s-%s-%s"
-	AWSPartition      = "aws"
-	AWSChinaPartition = "aws-cn"
+	snsArnPattern = "arn:%s:sns:%s:%s:%s-%s-%s-%s"
```
Since "arn" is an acronym and unique name we should use either all cap or all lower-case, and all cap in this case.

We haven't been perfectly consistent on this in history, but I think we've started to adapt to all-lower/all-cap at some point. 
```suggestion
	snsARNPattern = "arn:%s:sns:%s:%s:%s-%s-%s-%s"
```

Lou1415926 commented on internal/pkg/template/workload.go:
```diff
@@ -440,7 +469,20 @@ func envControllerParameters(o WorkloadOpts) []string {
 }
+
+// URI determines the uri for a queue using the queue name and account information
+func (q SQSQueue) URI(topicName *string) string {
```
One advantage of using pointer type as the receiver is to avoid copying the values on each call. In this case it's fine since `SQSQueue` is not a large struct 😉. But we should keep this in mind when determining the receiver type!

Lou1415926 commented on internal/pkg/template/workload.go:
```diff
@@ -440,7 +469,20 @@ func envControllerParameters(o WorkloadOpts) []string {
+	} else if topicName != nil {
+		name = aws.StringValue(topicName)
+	} else {
+		name = "SQSQueue"
```
nit: We can maybe define a `const` for this on top of the file
```go
const defaultQueueName = "SQSQueue"
```

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -473,17 +481,33 @@ func validateTopicSubscription(ts manifest.TopicSubscription, validTopicARNs []s
-			continue
+		arn, err := arn.Parse(topicARN)
+		if err != nil {
+			return err
```
Should we perhaps just `continue` and match `topicName` against the next `validTopicName` here?

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/validate.go:
```diff
@@ -463,7 +471,7 @@ func isCorrectSvcNameFormat(s string) bool {
 }
 
-func validateTopicSubscription(ts manifest.TopicSubscription, validTopicARNs []string) error {
+func validateTopicSubscription(ts manifest.TopicSubscription, validTopicARNs []string, app, env string) error {
```
A question here 🙋‍♀️:

When would the `app` and `env` parameters passed into this function be different from the "<app>" and "<env>" component of the valid topic names?

xar-tol replied:
If someone tries to add something weird in the manifest. For example, the wanted topic name may have "svc: svc-topicname" and name "bad topic", while the valid topic would have "svc: svc" "name: topic name". This would be a security concern, as the wanted topic needs to match a current valid topic exactly.

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/validate_test.go:
```diff
@@ -567,7 +569,9 @@ func TestValidateWorkerName(t *testing.T) {
+	validTopics := []string{"arn:aws:sns:us-east-1:123456789012:app-env-svc-name", "arn:aws:sns:us-east-1:123456789012:app-env-svc-name2"}
+	app := "app"
+	env := "env"
 	testCases := map[string]struct {
```
We can add one more test case that shows how we behave when one of the `validTopics` cannot be parsed.

xar-tol committed:
adjusts duration usage

Lou1415926 commented on internal/pkg/template/workload.go:
```diff
@@ -440,7 +469,20 @@ func envControllerParameters(o WorkloadOpts) []string {
+	} else {
+		name = "SQSQueue"
+	}
+	return fmt.Sprintf(sqsURIPattern, q.URL, q.AccountID, q.App, q.Env, q.Svc, name)
```
Will this URL be constructable in the template? If yes, this may probably alleviate the parameter drilling (i.e. passing parameters deep down only because the functions at the end need it) that we have right now.

bvtujo replied:
Yes and no. We need to pass the queue URLs into a JSON environment variable, which is difficult. The other reason is security--for sns topics, we need to validate that the constructed topic name for the subscription (with app, env, service, and topic name) exactly matches one of the topics previously deployed by copilot. But otherwise policies or resources which need queue ARNs or URLs as inputs can get them by calling !Ref.

Lou1415926 reviewed (commented):
Amazing work! Such complex logic and clear break down of PRs. 

I have some thoughts and questions below ⬇️

xar-tol committed:
adds subscribe tests to worker_svc

xar-tol committed:
adjusts template bools to non ptr

adjusts bools in template to be non-pointer due to CF issues

xar-tol committed:
removes unused var

Lou1415926 commented on internal/pkg/deploy/cloudformation/stack/transformers.go:
```diff
@@ -711,7 +711,7 @@ func convertTopicQueue(q *manifest.SQSQueue, url, accountID, app, env, svc strin
 		Delay:      delay,
 		Timeout:    timeout,
-		KMS:        q.KMS,
+		KMS:        aws.BoolValue(q.KMS),
```
Sorry this is not part of the PR but here is a question ⬇️ 

https://github.com/aws/copilot-cli/blob/e05a5b592e6177cfddb5c1cb04855091535826fe/internal/pkg/manifest/worker_svc.go#L71

Is it because we want to leave room for accepting a custom KMS key (thus a string, not a bool) in the future that we name it `kms` instead of `encrypted` in the manifest?

xar-tol replied:
The current plan is for this to create a new kms key just for that queue, and that encryption is always on by default. Thus, 'encrypted' would have been misleading.

xar-tol replied:
We decided to remove KMS entirely and do it by default - so this no longer applies.

Lou1415926 commented on internal/pkg/template/workload.go:
```diff
@@ -471,18 +474,18 @@ func envControllerParameters(o WorkloadOpts) []string {
 }
 
-// URI determines the uri for a queue using the queue name and account information
+// URI determines the uri for a queue using the queue name, topic name, and account information
```
```suggestion
// URI determines the uri for a queue using the queue name or topic name, and account information
```

Lou1415926 commented on internal/pkg/manifest/worker_svc_test.go:
```diff
@@ -410,6 +410,127 @@ func TestWorkerSvc_ApplyEnv(t *testing.T) {
+			Type: aws.String(WorkerServiceType),
+		},
+		WorkerServiceConfig: WorkerServiceConfig{
+			Subscribe: &SubscribeConfig{},
```
No need to update your code - just thinking out loud! I wonder what manifest input will give us `&SubscribeConfig{}` for `Subscribe` 🤔  If they don't have `subscribe` in manifest at all i'm assuming it'll just be `nil` (please correct me if this is wrong!).

xar-tol replied:
Good point, it looks like it's nil, so I removed that test case since nil is already covered.

Lou1415926 commented on internal/pkg/manifest/worker_svc_test.go:
```diff
@@ -410,6 +410,127 @@ func TestWorkerSvc_ApplyEnv(t *testing.T) {
 		},
 	}
+	duration111Seconds := 111 * time.Second
+	mockWorkerServiceWithSubscribeNilOverride := WorkerService{
```
I feel like it may be worthwhile to also add tests specifically for `Topics` field. Manifest field of type list is always the ones that' are error-prone when it comes to environment override 😭. 

We have had issues with list fields many times (one PR that fixes an issue with a list of security groups: https://github.com/aws/copilot-cli/pull/2442, another PR that fixes an issue with entrypoint which is a list of strings: https://github.com/aws/copilot-cli/pull/2490)

I think since you are using `*[]string` instead of `[]string` it should be fine. We can just add the unit tests to make sure.

xar-tol replied:
Added!

Lou1415926 reviewed (approved):
Looking great!!! Nothing major - just two questions and a comment on unit test!

xar-tol committed:
removes kms, adjusts DL ID to name, adds topic tests for worker

xar-tol committed:
removes account info and names from queues

removes act info from queues due to new json generation
removes names from queues to prevent replacement errors

xar-tol committed:
removes unneeded uri pattern

bvtujo commented on internal/pkg/deploy/cloudformation/stack/transformers_test.go:
```diff
@@ -1585,28 +1585,26 @@ func Test_convertSubscribe(t *testing.T) {
 						Service: "svc",
 					},
 				},
 				Queue: &manifest.SQSQueue{
```
Can we add a test for an empty (or minimal) queue field in the manifest and make sure that the template object  gets populated with the default values appropriately?

bvtujo reviewed (approved):
Looks amazing. Such thorough testing and so well constructed!

xar-tol committed:
adds tests for queues

mergify[bot] committed:
Merge branch 'mainline' into xar-tol/adds-queue-to-worker
